core/runtime/common/runtime_upgrade_tracker_impl.cpp

```diff
@@ -19,124 +19,109 @@ namespace kagome::runtime {
     BOOST_ASSERT(header_repo_);
   }
 
+  bool RuntimeUpgradeTrackerImpl::isStateInChain(
+      const primitives::BlockInfo &state,
+      const primitives::BlockInfo &chain_end) const {
+    // if the found state is finalized, it is guaranteed to not belong to a
+    // different fork
+    if (block_tree_->getLastFinalized().number >= state.number) {
+      return true;
+    }
+    // a non-finalized state may belong to a different fork, need to check
+    // explicitly (can be expensive if blocks are far apart)
+    KAGOME_PROFILE_START(has_direct_chain)
+    bool has_direct_chain =
+        block_tree_->hasDirectChain(state.hash, chain_end.hash);
+    KAGOME_PROFILE_END(has_direct_chain)
+    return has_direct_chain;
+  }
+
+  outcome::result<boost::optional<storage::trie::RootHash>>
+  RuntimeUpgradeTrackerImpl::findProperFork(
+      const primitives::BlockInfo &block,
+      std::vector<RuntimeUpgradeData>::const_iterator latest_state_update_it)
+      const {
+    for (; std::next(latest_state_update_it) != runtime_upgrades_.begin();
```

Comment:
What if we are not finding proper runtime block among forks? This condition would lead up to the runtime_upgrades stack beginning.

Reply:
No point to iterate over finalized block runtimes

Reply:
Wait, we don't. isStateInChain immediately returns true for finalised runtimes.

Reply:
If we never find a proper upgrade, we will stop at std::next(latest_state_update_it) != runtime_upgrades_.begin(), meaning that latest_state_update_it was pointing to begin() on the previous iteration, then after decrement it became begin()-1 and now next(begin()-1) == begin(). We return boost::none and then proceed to the orphan block case in the getLastCodeUpdateState.

Reply:
yeah, right, it is ok